pkg/controllers/managementuser/resourcequota/resource_quota_sync.go

```diff
@@ -385,42 +387,29 @@ func (c *SyncController) setValidated(ns *corev1.Namespace, value bool, msg stri
 	return c.Namespaces.Update(toUpdate)
 }
 
-func (c *SyncController) getResourceQuotaToUpdate(ns *corev1.Namespace) (string, error) {
-	quota := getNamespaceResourceQuota(ns)
+func (c *SyncController) getResourceQuotaToUpdate(ns *corev1.Namespace, existingQuota *v32.NamespaceResourceQuota) (*v32.NamespaceResourceQuota, error) {
 	defaultQuota, err := getProjectNamespaceDefaultQuota(ns, c.ProjectLister)
 	if err != nil {
-		return "", err
+		return nil, err
+	}
+	if existingQuota == nil {
+		return defaultQuota, nil
 	}
 
 	// rework after api framework change is done
 	// when annotation field is passed as null, the annotation should be removed
 	// instead of being updated with the null value
 	var updatedQuota *v32.NamespaceResourceQuota
-	if quota != "" && quota != "null" {
+	if existingQuota != nil {
 		// check if fields need to be removed or set
 		// based on the default quota
-		var existingQuota v32.NamespaceResourceQuota
-		err := json.Unmarshal([]byte(convert.ToString(quota)), &existingQuota)
-		if err != nil {
-			return "", err
-		}
-		updatedQuota, err = completeQuota(&existingQuota, defaultQuota)
-		if updatedQuota == nil || err != nil {
-			return "", err
-		}
-	}
-
-	var b []byte
-	if updatedQuota == nil {
-		b, err = json.Marshal(defaultQuota)
-	} else {
-		b, err = json.Marshal(updatedQuota)
+		updatedQuota, err = completeQuota(existingQuota, defaultQuota)
```

Comment:
In the UI, you can't specify namespace limits for things not specified on the project.
In kubectl, you can specify, for example, configMaps via the annotation, and then they become part of the quota limit (if the project does not set a limit on configMaps).

Reply:
If the UI prevents you from setting namespace limits for resources which have no limits set in the project than this conversation is more academic than anything else.

That being said, I don' think that a namespace specific limit for a newer type will be included. CompleteQuota won't include these limits unless something is defined in default quota (from the project) that it can override because of how the looping structure works. E.x. below:

1. Create a project - don't define a resourceLimit for configMaps.
2. Create a namespace, define a resourceLimit for configMaps (e.g. 50).
3. completeQuota is called.
4. Since there is no limit for configMaps in the project, no limit for configMaps is added to the combined map.
5. The namespace does not get a resourceLimit for configMaps despite explicitly asking for it.

Thoughts?

Reply:
That's the thing - if you try this via kubectl, then the namespace does get this resource limit for configMaps, when my branch's code is run.

Reply:
Talked about this yesterday - conclusion was that this is working because the memory that the function is using to store the computed map is pre-allocated with a deepcopy from the project limit, which causes this merge to happen. Resolving since this was fixed in the latest commits.